### ISSUE TITLE
Numeric settings with 'metadata=True' give metadatactrl errors

### ISSUE THREAD

**Issue by votti** [CellProfiler/CellProfiler#4276]
**Describe the bug**
I noticed that setting 'metadata=True' for a numeric setting (e.g. `.setting.text.Integer`) yields errors akin to:

```
TypeError: object of type 'int' has no len()
Traceback (most recent call last):
  File "cellprofiler\gui\pipelinelistview.py", line 1069, in __on_item_activated

  File "cellprofiler\gui\module_view\_module_view.py", line 486, in set_selectio
n
  File "cellprofiler\gui\module_view\_module_view.py", line 1592, in make_text_c
ontrol
  File "cellprofiler\gui\metadatactrl.py", line 61, in __init__
  File "cellprofiler\gui\metadatactrl.py", line 125, in SetValue
TypeError: object of type 'int' has no len()
```

**To Reproduce**
Produce a custom Module with a `Integer` setting and `metadata=True`

This is the case e.g. in my custom module: https://github.com/BodenmillerGroup/ImcPluginsCP/blob/d559e27c1fb31450a13150550158f23538cb8bb1/plugins/cropimage.py#L77

**Expected behavior**
No error :)

**Desktop (please complete the following information):**
- CP4
- Windows and Ubuntu

**Additional context**

I think this is because the `make_text_control` method just uses the value of the setting but expects this to be a string:
https://github.com/CellProfiler/CellProfiler/blob/8b9ef26b3437ed558c9dde962bf42829ae49bb1a/cellprofiler/gui/module_view/_module_view.py#L1596

For numeric settings these will be numeric. Changing this line to coerce the value to string would fix this:
`value=str(v.value)`

However I am not sure if there are other parts of the code that have issues with non-numeric setting values + `metadata=True` and I don't know how to write tests including GUI things.

**Comment by DavidStirling** [comment]
Hi @votti,

You're right that the `Integer` setting type just runs through the generic text widget generator. I'd like to understand why you're trying to pass it `metadata=True` though? If I'm reading your code correctly are you trying to use metadata tags to specify that integer?

The `Integer` field type includes it's own validation to confirm that the entered text is a valid integer, but because of the way metadata tags work we don't have a system to validate inputs that would come from such tags, thus they're not typically available in these fields. The metadata tags themselves are primarily intended for generating pathnames. In theory you could potentially use a generic `Text` widget with metadata and validate/convert that to `int` within your module if you really want to do this.

To put it another way: metadata tags are arbitrary text, so I'm not sure that permitting them in fields which are explicitly for numeric values would make a lot of sense. Is there an intended use-case for this?

**Comment by votti** [comment]
Hi,

I used this functionality previously to allow for cropping of a specific
sections of an image based on coordinates provided by metadata ( eg via a
metadata csv).

I guess it was working like this in CP3 and now just broke in CP4,  so I
considered it as a bug - but maybe I was just abusing this all along 😅

If this is really not intended use I would suggest to also add a validation
statement that captures the 'metadata' keyword at the 'Numeric' parent
class and throw an error to make this more explicit. Is this something I
should do?

Generally I did like the idea of being able to provide numeric values via
Metadata as it could make sense in other contexts as well (eg providing
image specific thresholds or so).
The lack of validation is technically true but then the feature is fairly
hidden so I doubt someone would use it without intend.

In any case for my module I could also just use a pure 'text' setting
instead and validate the settings manually.

Cheers!

**Comment by DavidStirling** [comment]
Thanks for elaborating, I can see where that might be useful!

The internal structure of CellProfiler has changed a lot from CP3, so I'm not surprised that this stopped working. I would say that this was never an intended use for the tags system, but if we coerce the input values into strings as suggested it could resolve the blocking error. However, that still won't play nicely with the Integer class's validator. We don't know the contents of a metadata tag until run time, so I think we'd have to disable the validation check entirely, at which point you might as well just use the base text class.

My main concern with doing this is potential knock-on effects on other modules, since a lot of different widgets use the text entry constructor. Would you be able to test if everything works as expected if using a standard Text setting? It should just be a case of calling `int()` on the value when you fetch it within the module.

Thanks!

**Comment by votti** [comment]
For my module moving to standard Text settings is a fine solution.

I will still add a check to the `__init__` of the `Numbers` setting to check that `metadata=True` is not passed as a kwarg and throw a `ValueError`. I will do a pull request shortly.

This will make this clear and explicit and crash the module at initialization instead to somewhere in the background only when running the GUI.

**Comment by DavidStirling** [comment]
Fixed by Cellprofiler/core#57, much appreciated!
